### PR TITLE
feat(generator): generate repository metadata files

### PR DESCRIPTION
**Pull request by coryan**
The documentation index is based on the `.repo-metadata.json` files
committed to our GitHub repository. With this change, the generator
will automatically create and update these files.

We need to get some information from the service config YAML file, so
this change introduces changes to find and parse the file. There are
other cool bits of information on that file, which we can use later.

The generator only creates a `.repo-metadata.json` file if all the
necessary information is found. This can be useful to manually create
the files when the information is missing, as the generator will not
overwrite them.

Part of the work for #11519

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11555)

### CODE REVIEW

File: generator/internal/scaffold_generator.cc [hunk 1]
```diff
@@ -138,6 +147,62 @@ std::map<std::string, std::string> ScaffoldVars(
...
+  // convert that to a link to list issues, which is what we want to generate.
+  auto l = vars.find("new_issue_uri");
+  if (l != vars.end()) {
+    auto issue_tracker = l->second;
+    auto const re = std::regex(
+        R"re((https://issuetracker.google.com/issues).*[^a-z]component=([0-9]*).*)re");
+    std::smatch match;
+    if (std::regex_match(issue_tracker, match, re)) {
+      issue_tracker = absl::StrCat(
+          match[1].str(), "?q=componentid:", match[2].str(), " status=open");
```
Comment (dbolduc) [inline generator/internal/scaffold_generator.cc:201]: `' '` in the link seems weird, but :shrug:

File: generator/internal/scaffold_generator.cc [hunk 2]
```diff
@@ -149,11 +214,77 @@ void MakeDirectory(std::string const& path) {
...
+    google::cloud::cpp::generator::ServiceConfiguration const& service,
+    bool allow_placeholders) {
+  auto const destination = [&]() {
+    MakeDirectory(output_path);
+    auto d = output_path + "/" + LibraryPath(service.product_path());
+    MakeDirectory(d);
+    auto l = vars.find("service_subdirectory");
+    if (l == vars.end()) return d;
+    if (l->second.empty()) return d;
+    d += l->second;
```
Comment (dbolduc) [inline generator/internal/scaffold_generator.cc:229]: will the subsequent `MakeDirectory(d)` call work if `l->second` is like `foo/v1` ?

File: generator/internal/scaffold_generator.cc [hunk 3]
```diff
@@ -149,11 +214,77 @@ void MakeDirectory(std::string const& path) {
...
+    d += "/";
```
Comment (dbolduc) [inline generator/internal/scaffold_generator.cc:230]: I don't think this is necessary. `service_subdirectory` should always end in a slash.

https://github.com/googleapis/google-cloud-cpp/blob/c435bc9ac0cc6dc26833164a3f28e82c111eced4/generator/internal/scaffold_generator.cc#L82

**Review by dbolduc — APPROVED** [review]

**Review by dbolduc — APPROVED** [review]

File: generator/internal/scaffold_generator.cc (hunk 3 above)
Comment (coryan) [reply generator/internal/scaffold_generator.cc]: Removed.

File: generator/internal/scaffold_generator.cc (hunk 2 above)
Comment (coryan) [reply generator/internal/scaffold_generator.cc:230]: Don't know, but that is what the rest of the scaffold does.

File: generator/internal/scaffold_generator.cc (hunk 1 above)
Comment (coryan) [reply generator/internal/scaffold_generator.cc]: Fixed

File: generator/generator_config.proto [hunk 4]
```diff
@@ -144,6 +144,10 @@ message ServiceConfiguration {
   // This need not be explicitly set to true as the generator will check if the
   // field has been set, and assign a value of true if not set.
   optional bool generate_grpc_transport = 22;
+
+  // If set to `true` the generator will *NOT* create a .repo-metadata.json
+  // file. This is useful when multiple services target the same directory.
+  bool omit_metadata = 23;
```
[the remark this answers is not in the document]
Comment (coryan) [reply generator/generator_config.proto]: Fixed

**Pull request merged**